// xero-node/lib/__integration_tests__/attachments.integration.tests.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var _this = this;
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var path = require("path");
var AccountingAPIClient_1 = require("../AccountingAPIClient");
var invoice_request_examples_1 = require("./request-body/invoice.request.examples");
var integration_helpers_1 = require("./helpers/integration.helpers");
describe('attachments', function () {
    var xero;
    var invoiceIdsToArchive = [];
    var anInvoice;
    var attachementPath = path.resolve(__dirname, 'helpers', 'image.jpg');
    var tempAttachementPath = path.resolve(__dirname, 'temp-image.jpg');
    beforeAll(function () { return __awaiter(_this, void 0, void 0, function () {
        var config, response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    integration_helpers_1.setJestTimeout();
                    config = integration_helpers_1.getPrivateConfig();
                    xero = new AccountingAPIClient_1.AccountingAPIClient(config);
                    return [4 /*yield*/, xero.invoices.create(invoice_request_examples_1.createSingleInvoiceRequest)];
                case 1:
                    response = _a.sent();
                    anInvoice = response.Invoices[0];
                    collectInvoicesToArchive(response);
                    return [2 /*return*/];
            }
        });
    }); });
    it('can add an attachment', function () { return __awaiter(_this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, xero.invoices.attachments.uploadAttachment({
                        entityID: anInvoice.InvoiceID,
                        mimeType: 'image/jpg',
                        fileName: 'anUploadedImage.jpg',
                        pathToUpload: attachementPath
                    })];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    }); });
    it('can download the same attachment', function () { return __awaiter(_this, void 0, void 0, function () {
        var tempFile, realFile;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, xero.invoices.attachments.downloadAttachment({
                        entityID: anInvoice.InvoiceID,
                        mimeType: 'image/jpg',
                        fileName: 'anUploadedImage.jpg',
                        pathToSave: tempAttachementPath
                    })];
                case 1:
                    _a.sent();
                    tempFile = fs.statSync(tempAttachementPath);
                    realFile = fs.statSync(attachementPath);
                    expect(tempFile.size).toBe(realFile.size);
                    return [2 /*return*/];
            }
        });
    }); });
    it('get the attachment details', function () { return __awaiter(_this, void 0, void 0, function () {
        var response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, xero.invoices.attachments.get({
                        EntityID: anInvoice.InvoiceID
                    })];
                case 1:
                    response = _a.sent();
                    expect(response.Attachments.length).toEqual(1);
                    expect(response.Attachments[0].FileName).toEqual('anUploadedImage.jpg');
                    return [2 /*return*/];
            }
        });
    }); });
    it('errors like this', function () { return __awaiter(_this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/];
        });
    }); });
    afterAll(function () { return __awaiter(_this, void 0, void 0, function () {
        var updateRequestBody;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    // delete the file
                    fs.unlinkSync(tempAttachementPath);
                    updateRequestBody = invoiceIdsToArchive.map(function (invoiceId) { return ({ InvoiceID: invoiceId, Status: 'DELETED' }); });
                    return [4 /*yield*/, xero.invoices.update({ Invoices: updateRequestBody })];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    }); });
    function collectInvoicesToArchive(response) {
        invoiceIdsToArchive = invoiceIdsToArchive.concat(response.Invoices.map(function (invoice) { return invoice.InvoiceID; }));
    }
});
