// xero-node/lib/__integration_tests__/prepayments.integration.tests.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var _this = this;
Object.defineProperty(exports, "__esModule", { value: true });
var AccountingAPIClient_1 = require("../AccountingAPIClient");
var integration_helpers_1 = require("./helpers/integration.helpers");
describe.skip('/prepayments', function () {
    var xero;
    var existingId;
    beforeAll(function () {
        integration_helpers_1.setJestTimeout();
        var config = integration_helpers_1.getPrivateConfig();
        xero = new AccountingAPIClient_1.AccountingAPIClient(config);
    });
    it('create via banktransactions endpoint', function () { return __awaiter(_this, void 0, void 0, function () {
        var response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, xero.banktransactions.create({
                        Type: 'RECEIVE-PREPAYMENT',
                        Contact: { ContactID: '6d42f03b-181f-43e3-93fb-2025c012de92' },
                        BankAccount: { Code: '090' },
                        LineAmountTypes: 'Exclusive',
                        LineItems: [
                            {
                                Description: 'Prepayment for Kitchen Designs',
                                Quantity: 1,
                                UnitAmount: 500.00,
                                AccountCode: '200'
                            }, {
                                Description: 'Prepayment for Kitchen materials',
                                Quantity: 1,
                                UnitAmount: 1000.00,
                                AccountCode: '200'
                            }
                        ]
                    })];
                case 1:
                    response = _a.sent();
                    expect(response.Payments.length).toBe(1);
                    expect(response.Payments[0].Code).toEqual('090');
                    existingId = response.Payments[0].PaymentID;
                    return [2 /*return*/];
            }
        });
    }); });
    it('get all', function () { return __awaiter(_this, void 0, void 0, function () {
        var response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, xero.prepayments.get()];
                case 1:
                    response = _a.sent();
                    expect(response).toBeDefined();
                    expect(response.Id).toBeTruthy();
                    expect(response.Prepayments.length).toBeGreaterThan(0);
                    expect(response.Prepayments[0].PrepaymentID).toBeTruthy();
                    return [2 /*return*/];
            }
        });
    }); });
    it('get single', function () { return __awaiter(_this, void 0, void 0, function () {
        var response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, xero.prepayments.get({ PrepaymentID: existingId })];
                case 1:
                    response = _a.sent();
                    expect(response).toBeDefined();
                    expect(response.Id).toBeTruthy();
                    expect(response.Prepayments.length).toBe(1);
                    expect(response.Prepayments[0].PrepaymentID).toEqual(existingId);
                    return [2 /*return*/];
            }
        });
    }); });
    it('reverse via the payments endpoint', function () { return __awaiter(_this, void 0, void 0, function () {
        var response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, xero.payments.update({ PaymentID: existingId, Status: 'DELETED' })];
                case 1:
                    response = _a.sent();
                    expect(response).toBeDefined();
                    expect(response.Id).toBeTruthy();
                    expect(response.Payments.length).toBe(1);
                    expect(response.Payments[0].PaymentID).toEqual(existingId);
                    expect(response.Payments[0].Status).toEqual('DELETED');
                    return [2 /*return*/];
            }
        });
    }); });
});
